crankshaft-engine/src/service/runner/backend/docker.rs

```diff
@@ -59,67 +308,123 @@ impl crate::Backend for Backend {
+                    if let Some(work_dir) = execution.work_dir() {
+                        builder = builder.work_dir(work_dir);
+                    }
 
-                // (4) Cleanup the container (if desired).
-                if cleanup {
-                    container
-                        .remove()
-                        .await
-                        .wrap_err("failed to remove container")?
+                    let service = Arc::new(builder.try_build(&name).await?);
+
+                    let result = select! {
+                        _ = token.cancelled() => {
+                            Err(eyre!("task has been cancelled"))
+                        }
+                        res = Self::run_with_service(execution_index, started.clone(), service.clone()) => {
+                            res
+                        }
+                    };
+
+                    if cleanup {
+                        service
+                            .delete()
+                            .await
+                            .wrap_err("failed to delete service")?
+                    }
+
+                    outputs.push(result?);
                 }
+            } else {
+                let name = task
+                    .name()
+                    .context("task requires a name to run on the Docker backend")?;
+
+                let config = HostConfig {
+                    mounts: Some(mounts.clone()),
+                    ..task.resources().map(HostConfig::from).unwrap_or_default()
+                };
 
-                outputs.push(output);
+                for (execution_index, execution) in task.executions().enumerate() {
+                    if token.is_cancelled() {
+                        bail!("task has been cancelled");
+                    }
+
+                    let mut builder = client
+                        .container_builder()
+                        .image(execution.image())
+                        .program(execution.program())
+                        .args(execution.args())
+                        .attach_stdout()
+                        .attach_stderr()
+                        .host_config(config.clone());
+
+                    if let Some(work_dir) = execution.work_dir() {
+                        builder = builder.work_dir(work_dir);
+                    }
+
+                    let container = Arc::new(
+                        builder
+                            .try_build(&name)
+                            .await?,
+                    );
+
+                    let result = select! {
+                        _ = token.cancelled() => {
+                            Err(eyre!("task has been cancelled"))
+                        }
+                        res = Self::run_with_container(&task, execution_index, started.clone(), container.clone()) => {
+                            res
+                        }
+                    };
+
+                    if cleanup {
```

Comment:
Should this cleanup just happen after the `if`/`else` statement and be uniformly handled across either service/non-service dispatches?